homeassistant/util/logging.py

```diff
@@ -2,21 +2,93 @@
 
 from __future__ import annotations
 
+from collections import defaultdict
 from collections.abc import Callable, Coroutine
+import datetime
 from functools import partial, wraps
 import inspect
 import logging
 import logging.handlers
-import queue
+from queue import SimpleQueue
 import traceback
-from typing import Any, cast, overload
+from typing import Any, cast, overload, override
 
 from homeassistant.core import (
     HassJobType,
     HomeAssistant,
     callback,
     get_hassjob_callable_job_type,
 )
+from homeassistant.helpers.event import async_track_time_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HomeAssistantQueueListener(logging.handlers.QueueListener):
+    """Custom QueueListener to watch for noisy loggers."""
+
+    LOG_COUNTS_RESET_INTERVAL = datetime.timedelta(minutes=5)
+    MAX_LOGS_COUNT = 200
+
+    _reset_pending: bool
+    _log_counts: dict[str, int]
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        queue: SimpleQueue[logging.Handler],
+        *handlers: logging.Handler,
+    ) -> None:
+        """Initialize the handler."""
+        super().__init__(queue, *handlers)
+        self._hass = hass
+        self._reset_counters()
+
+        @callback
+        def _set_reset_pending_flag(_: datetime.datetime) -> None:
+            _LOGGER.debug("Setting reset pending")
+            # Set a boolean instead of directly resetting the counters so that the actual
+            # reset happens on the QueueListener thread without the need for a Lock.
+            self._reset_pending = True
```

Comment:
I was divided on weather or not to just reset the dict here, since it is thread safe anyway, but I think it is safer to do it on the queue thread since we read and write to it separately. that would not cause an issue right now, but makes introducing a bug easier in the future.